library/internal/buffer.go

```diff
@@ -0,0 +1,64 @@
+package internal
+
+import (
+	"bytes"
+	"io"
+)
+
+//MultiplexingBuffer is a type that will allow storing, replaying and streaming data from it's internal buffer
+//It implements io.Writer storing the written bytes to an unbound buffer internally
+//The Reader function will return an io.Reader that will stream data starting with the buffer, until the Close method of the underlying MultiplexingBuffer is called
+type MultiplexingBuffer struct {
+	closed bool
+	buffer []byte
+}
+
+type streamingReader struct {
+	multiplexingBuffer *MultiplexingBuffer
+	idx                int
+	drained            bool
+}
+
+func (s *streamingReader) Read(p []byte) (n int, err error) {
```

Comment:
Sorry it's not very clear to me - how are clients that are streaming logs notified about whether there's new output or not?

Reply:
I tried to follow the logic of io.Reader where on encountering an end of file (which is in our case the end of the stream) the reader first returns the number of files read, and the subsequently 0, EOF on all calls